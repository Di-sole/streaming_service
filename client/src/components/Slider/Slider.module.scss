.container {
    display: flex;
    position: relative;
    height: 100%;

    &_preview,
    &_rating,
    &_actor {
        margin: 0 -12px;

        .container__contentContainer {
            padding: 15px 0 0 0;
        }

        .container__button {
            &_left {
                left: -40px;
            }

            &_right {
                right: -40px;
            }
        }
    }

    &_preview {
        .container__button {
            &_left {
                img {
                    top: 46%;
                }
            }

            &_right {
                img {
                    top: 46%;
                }
            }
        }
    }

    &_actor {
        .container__button {
            &_left {
                img {
                    top: 42%;
                }
            }

            &_right {
                img {
                    top: 42%;
                }
            }
        }
    }

    &__contentContainer {
        width: 100%;
        overflow: hidden;
    }

    &__content {
        position: relative;
        transition: 0.4s;
        display: flex;
        width: 100%;
    }

    &__item {
        min-width: calc(100% / 7 - 10px);
        min-height: calc(100% / 7 - 10px * 3);
        margin: 0 5px;
    }

    &__button {
        width: 60px;
        height: 95%;
        position: absolute;
        top: 0;
        margin: 10px 0 30px 0;

        &_left {
            left: -50px;
        }

        &_right {
            right: -50px;
        }
    }
}

@media (max-width: 1380px) {
    .container {
        &__button {
            width: 22px;

            &_left {
                left: -22px !important;
            }

            &_right {
                right: -22px !important;
            }
        }

        &_preview,
        &_rating,
        &_actor {
            .container__button {
                width: 22px;

                &_left {
                    left: -10px !important;
                }

                &_right {
                    right: -10px !important;
                }
            }
        }
    }
}
/*
@media (max-width: 600px) {
    .container {
        &__content {
            &_preview {
                max-height: 210px;
            }
        }
    }
}

@media (max-width: 512px) {
    .container {
        &__content {
            &_preview {
                max-height: 260px;
            }
        }
    }
}

@media (max-width: 400px) {
    .container {
        &__content {
            &_preview {
                max-height: 300px;
            }
        }
    }
}
*/
