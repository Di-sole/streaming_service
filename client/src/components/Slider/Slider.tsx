import React, { ReactNode, useEffect, useState } from 'react';
import styles from './Slider.module.scss';
import ArrowButton from '../UI/buttons/ArrowButton/ArrowButtonUI';
import { useWindowSize } from '@/hooks/useWindowSize/useWindowSize';
import { useMoviesCount } from '@/hooks/useMoviesCount/useMoviesCount';
import { useRatingCount } from '@/hooks/useRatingCount/useRatingCount';

interface SliderProps {
    itemType: 'actor' | 'preview' | 'rating' | 'promo' | 'other';
    children: ReactNode;
    length: number;
}

export const Slider = ({ itemType, children, length }: SliderProps) => {
    const windowWidth = useWindowSize();
    const [position, setPosition] = useState<number>(0);

    const listCount =
        itemType === 'actor' || itemType === 'preview'
            ? useMoviesCount(windowWidth)
            : itemType === 'rating'
            ? useRatingCount(windowWidth)
            : itemType === 'promo'
            ? 1
            : 2;

    const minPosition = 0;
    const maxPosition = (-100 / listCount) * length + 100;

    const leftHandler = () => {
        const newPosition = itemType !== 'promo' ? position + (100 / listCount) * (listCount - 1) : position + 100;

        if (newPosition > minPosition) {
            setPosition(minPosition);
        } else {
            setPosition(newPosition);
        }
    };

    const rightHandler = () => {
        const newPosition = itemType !== 'promo' ? position - (100 / listCount) * (listCount - 1) : position - 100;

        if (newPosition < maxPosition) {
            setPosition(maxPosition);
        } else {
            setPosition(newPosition);
        }
    };

    useEffect(() => {
        setPosition(minPosition);
    }, [listCount]);

    return (
        <>
            <div className={[styles.container, styles[`container_${itemType}`]].join(' ')} data-testid={'slider'}>
                {position < minPosition && (
                    <ArrowButton
                        className={[styles.container__button, styles.container__button_left].join(' ')}
                        direction="left"
                        iconSize={windowWidth > 1380 ? 'large' : 'medium'}
                        onClick={leftHandler}
                    />
                )}

                <div className={[styles.container__contentContainer].join(' ')}>
                    <div className={[styles.container__content, 'content'].join(' ')}>{children}</div>
                </div>

                {position > maxPosition && (
                    <ArrowButton
                        className={[styles.container__button, styles.container__button_right].join(' ')}
                        direction="right"
                        iconSize={windowWidth > 1380 ? 'large' : 'medium'}
                        onClick={rightHandler}
                    />
                )}
            </div>

            <style jsx>
                {`
                    .content {
                        left: calc(0% + ${position}%);
                    }
                `}
            </style>
        </>
    );
};

export default Slider;
