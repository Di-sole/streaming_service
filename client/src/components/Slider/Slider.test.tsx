import { render, screen } from '@testing-library/react';
import Slider from './Slider';
import Image from 'next/image';

const Img = <Image src={require('../../../testAsserts/img/BG554460.jpg')} alt="Img" width={25} height={25} />;
describe('SLIDER SNAPSHOTS TESTS', () => {
    test('Slider should not be changed', () => {
        render(
            <Slider
                itemType="other"
                children={
                    <div>
                        {Img}
                        {Img}
                        {Img}
                        {Img}
                        {Img}
                        {Img}
                    </div>
                }
                length={5}
            />,
        );

        const id = screen.getByTestId('slider');
        expect(id).toMatchSnapshot();
    });
});
