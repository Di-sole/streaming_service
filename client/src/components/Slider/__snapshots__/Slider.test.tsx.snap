// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SLIDER SNAPSHOTS TESTS Slider should not be changed 1`] = `
<div
  class="jsx-687504307 container container_other"
  data-testid="slider"
>
  <div
    class="jsx-687504307 container__contentContainer"
  >
    <div
      class="jsx-687504307 container__content content"
    >
      <div>
        <img
          alt="Img"
          data-nimg="1"
          decoding="async"
          height="25"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
          style="color: transparent;"
          width="25"
        />
        <img
          alt="Img"
          data-nimg="1"
          decoding="async"
          height="25"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
          style="color: transparent;"
          width="25"
        />
        <img
          alt="Img"
          data-nimg="1"
          decoding="async"
          height="25"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
          style="color: transparent;"
          width="25"
        />
        <img
          alt="Img"
          data-nimg="1"
          decoding="async"
          height="25"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
          style="color: transparent;"
          width="25"
        />
        <img
          alt="Img"
          data-nimg="1"
          decoding="async"
          height="25"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
          style="color: transparent;"
          width="25"
        />
        <img
          alt="Img"
          data-nimg="1"
          decoding="async"
          height="25"
          loading="lazy"
          src="/_next/image?url=%2Fimg.jpg&w=64&q=75"
          srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=64&q=75 2x"
          style="color: transparent;"
          width="25"
        />
      </div>
    </div>
  </div>
  <button
    class="button container__button container__button_right"
  >
    <img
      alt=""
      class="icon icon"
      data-nimg="1"
      decoding="async"
      height="40"
      loading="lazy"
      src="/_next/image?url=%2Fimg.jpg&w=48&q=75"
      srcset="/_next/image?url=%2Fimg.jpg&w=32&q=75 1x, /_next/image?url=%2Fimg.jpg&w=48&q=75 2x"
      style="color: transparent;"
      width="20"
    />
  </button>
</div>
`;
